atomicfu/src/appleMain/kotlin/kotlinx/atomicfu/locks/Synchronized.kt

```diff
@@ -0,0 +1,297 @@
+        } else {
+            if (!owner.compareAndSet(NO_OWNER, self)) {
+                waitForUnlockAndLock(self)
+            }
+        }
+    }
+
+    public fun tryLock(): Boolean {
+        var self = currentThreadId
+        if (self == 0L) {
+            currentThreadId = pthread_self().toLong()
+            self = currentThreadId
+        }
+        return if (owner.value == self) {
+            reEnterCount += 1
+            true
+        } else if (waiters.incrementAndGet() == 1 && owner.compareAndSet(NO_OWNER, self)) {
+            true
+        } else {
+            waiters.decrementAndGet()
+            false
+        }
+    }
+
+
+    private fun waitForUnlockAndLock(self: Long) {
+        withMonitor(monitor) {
+            while (!owner.compareAndSet(NO_OWNER, self)) {
+                monitor.waitWithDonation(owner.value)
+            }
+        }
+    }
+
+    public fun unlock() {
+        require (owner.value == currentThreadId)
+        if (reEnterCount > 0) {
+            reEnterCount -= 1
+        } else {
+            owner.value = NO_OWNER
+            if (waiters.decrementAndGet() > 0) {
+                withMonitor(monitor) {
+                    // We expect the highest priority thread to be woken up, but this should work
+                    // in any case.
+                    monitor.nativeMutex.notify()
+                }
+            }
+        }
+    }
+
+    private inline fun withMonitor(monitor: DonatingMonitor, block: () -> Unit) {
+        monitor.nativeMutex.enter()
+        return try {
+            block()
+        } finally {
+            monitor.nativeMutex.exit()
+        }
+    }
+
+    @OptIn(kotlin.experimental.ExperimentalNativeApi::class)
+    private class DonatingMonitor {
+        val nativeMutex = mutexPool.allocate()
+        val cleaner = createCleaner(nativeMutex) { mutexPool.release(it) }
+        var qosOverride: pthread_override_t? = null
+        var qosOverrideQosClass: UInt = 0U
+
+        fun waitWithDonation(lockOwner: Long) {
+            donateQos(lockOwner)
+            nativeMutex.wait()
+            clearDonation()
+        }
+
+        private fun donateQos(lockOwner: Long) {
+            if (lockOwner == NO_OWNER) {
+                return
+            }
+            val ourQosClass = qos_class_self() as UInt
+            // Set up a new override if required:
+            if (qosOverride != null) {
+                // There is an existing override, but we need to go higher.
+                if (ourQosClass > qosOverrideQosClass) {
+                    pthread_override_qos_class_end_np(qosOverride)
+                    qosOverride = pthread_override_qos_class_start_np(lockOwner.toCPointer(), qos_class_self(), 0)
+                    qosOverrideQosClass = ourQosClass
+                }
+            } else {
+                // No existing override, check if we need to set one up.
+                memScoped {
+                    val lockOwnerQosClass = alloc<UIntVar>()
```

Comment:
It's pretty much likely that this allocation will be happen on each contention and it's a real `malloc`.

Is it possible to extract this `alloc` into `DonatingMonitor` level and free it within cleaner?

Reply:
Shouldn't this just be a stack allocation (via memScoped)? It's just used short-lived for getting the values back, perhaps I am holding it wrong?

Reply:
`memScoped` is still a real allocation, we have no stack allocation in K/N AFAIR

Reply:
Done